keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -237,6 +196,91 @@ class ResourceActuator(
     }
   }
 
+  /**
+   * [VersionedArtifactProvider] is a special [resource] sub-type. When a veto response sets
+   * [VetoResponse.vetoArtifact], the resource under evaluation is of type
+   * [VersionedArtifactProvider], and the version has never before been deployed successfully to an environment,
+   * disallow the desired artifact version from being deployed to
+   * the environment containing [resource]. This ensures that the environment will be fully restored to
+   * a prior good-state.
+   *
+   * If the version has previously been deployed successfully do not veto the artifact because
+   * we do not want to veto the artifact in the case of a bad config change or other downstream problem.
+   * Rolling back most likely will not help in these cases and it will probably cause confusion.
+   */
+  private fun handleVetoingArtifact(
+    response: VetoResponse,
+    resource: Resource<*>,
+    deliveryConfig: DeliveryConfig,
+    environment: Environment,
+    desired: Any?
+  ) {
+    if (response.vetoArtifact && resource.spec is VersionedArtifactProvider) {
+      try {
+        val versionedArtifact = desired.findArtifact()
+
+        if (versionedArtifact != null) {
+          val artifact = deliveryConfig.matchingArtifactByName(versionedArtifact.artifactName, versionedArtifact.artifactType)
+            ?: error("Artifact ${versionedArtifact.artifactType}:${versionedArtifact.artifactName} not found in delivery config ${deliveryConfig.name}")
+
+          val previousSuccess = artifactRepository.wasSuccessfullyDeployedTo(
+            deliveryConfig = deliveryConfig,
+            artifact = artifact,
+            version = versionedArtifact.artifactVersion,
+            targetEnvironment = environment.name
+          )
+
+          if (!previousSuccess) {
+            artifactRepository.markAsVetoedIn(
+              deliveryConfig = deliveryConfig,
+              veto = EnvironmentArtifactVeto(
+                reference = artifact.reference,
+                version = versionedArtifact.artifactVersion,
+                targetEnvironment = environment.name,
+                vetoedBy = "Spinnaker",
+                comment = "Automatically marked as bad because multiple deployments of this version failed and none have ever succeeded."
+              )
+            )
+            log.info(
+              "Vetoing artifact version {} of artifact {} for config {} and env {} because multiple deploys failed",
+              versionedArtifact,
+              artifact.reference,
+              deliveryConfig.name,
+              environment.name
+            )
+            publisher.publishEvent(ArtifactVersionVetoed(artifact.reference))
+          } else {
+            log.info(
```

Comment:
Can you elaborate on how do you know, by `versionedArtifact == null`, that this artifact version was deployed successfully before?

Reply:
oh i'm sorry, I was confused with `previousSuccess==null`. nevermind :)